networks/hamiltonian_net.py

```diff
@@ -0,0 +1,116 @@
+            assert len(kernel_sizes) == hidden_conv_layers + 2 and \
+                   len(strides) == hidden_conv_layers + 2, \
+                   'kernel_sizes and strides must be lists with values foreach layer in the ' \
+                   'network (' + str(hidden_conv_layers + 2) + ' in this case).'
+        else:
+            raise ValueError('Args hidden_conv_layers, n_filters, kernel_sizes, and strides'
+                             'can only be either all None, or all defined by the user.')
+        in_channels = in_shape[0] * 2
+        paddings = [int(k/2) for k in kernel_sizes]
+        self.in_conv = nn.Conv2d(
+            in_channels=in_channels,
+            out_channels=n_filters[0],
+            kernel_size=kernel_sizes[0],
+            padding=paddings[0])
+        out_size = int((in_shape[1] - kernel_sizes[0] + 2 * paddings[0])) / strides[0] + 1
+        self.hidden_layers = nn.ModuleList(
+            modules=[
+                nn.Conv2d(
+                    in_channels=n_filters[i],
+                    out_channels=n_filters[i + 1],
+                    kernel_size=kernel_sizes[i + 1],
+                    padding=paddings[i + 1],
+                    stride=strides[i + 1]
+                )
+                for i in range(hidden_conv_layers)
+            ]
+        )
+        for i in range(hidden_conv_layers):
+            out_size = int((out_size - kernel_sizes[i] + 2 * paddings[i]) / strides[i]) + 1
+        self.out_conv = nn.Conv2d(
+            in_channels=n_filters[-1],
+            out_channels=n_filters[-1],
+            kernel_size=kernel_sizes[-1],
+            padding=paddings[-1]
+        )
+        out_size = int((out_size - kernel_sizes[-1] + 2 * paddings[-1]) / strides[-1]) + 1
+        self.n_flat = out_size ** 2 * n_filters[-1]
+        self.linear = nn.Linear(in_features=self.n_flat, out_features=1)
+        self.activation = act_func
+        self.type(dtype)
+
+    def forward(self, q, p):
+        """Forward pass that returns the Hamiltonian for the given q and p inputs.
+
+        q and p must be two (batch_size, channels, height, width) tensors.
+
+        Args:
+            q (torch.Tensor): The tensor corresponding to the position in abstract space.
+            p (torch.Tensor): The tensor corresponding to the momentum in abstract space.
+
+        Returns:
+            A (batch_size, 1) shaped tensor with the energy for each input in the batch.
+        """
+        x = torch.cat((q, p), dim=1)  # Concatenate q and p to obtain a N x 2C x H x W tensor
+        x = self.activation(self.in_conv(x))
+        for layer in self.hidden_layers:
+            x = self.activation(layer(x))
+        x = self.activation(self.out_conv(x))
+        x = x.view(-1, self.n_flat)
+        x = self.linear(x)
+        return x
+
+
+if __name__ == '__main__':
+    hamiltonian_net = HamiltonianNet(in_shape=(16, 4, 4))
+    q, p = torch.randn((2, 128, 16, 4, 4))
```

Comment:
probably im confused, but isn't there an extra dimension here?

Reply:
It is a dirty trick to avoid writing two lines. The first dimension = 2  is because `q` and `p` are two tensors, so the `(2, 128, 16, 4, 4)` tensor will be split in two and assigned to `q` and `p`. It is equivalent to writing:
```python
q = torch.randn((128, 16, 4, 4))
p = torch.randn((128, 16, 4, 4))
```